crates/core/src/json.rs

```diff
@@ -0,0 +1,281 @@
+extern crate alloc;
+
+use alloc::string::{String, ToString};
+use core::ffi::c_int;
+use serde::de::{DeserializeSeed, Deserializer, IgnoredAny, Visitor};
+
+use sqlite::ResultCode;
+use sqlite_nostd::{self as sqlite, ColumnType};
+use sqlite_nostd::{Connection, Context, Value};
+
+use crate::error::SQLiteError;
+use crate::util::context_set_error;
+use crate::{bson, create_sqlite_text_fn};
+
+/// Given any number of JSON TEXT arguments, merge them into a single JSON object.
+///
+/// This assumes each argument is a valid JSON object, with no duplicate keys.
+/// No JSON parsing or validation is performed - this performs simple string concatenation.
+fn powersync_json_merge_impl(
+    _ctx: *mut sqlite::context,
+    args: &[*mut sqlite::value],
+) -> Result<String, SQLiteError> {
+    if args.is_empty() {
+        return Ok("{}".to_string());
+    }
+    let mut result = String::from("{");
+    for arg in args {
+        let chunk = arg.text();
+        if chunk.is_empty() || !chunk.starts_with('{') || !chunk.ends_with('}') {
+            return Err(SQLiteError::from(ResultCode::MISMATCH));
+        }
+
+        // Strip outer braces
+        let inner = &chunk[1..(chunk.len() - 1)];
+
+        // If this is not the first chunk, insert a comma
+        if result.len() > 1 {
+            result.push(',');
+        }
+
+        // Append the inner content
+        result.push_str(inner);
+    }
+
+    // Close the outer brace
+    result.push('}');
+    Ok(result)
+}
+
+create_sqlite_text_fn!(
+    powersync_json_merge,
+    powersync_json_merge_impl,
+    "powersync_json_merge"
+);
+
+/// A variant of `json_extract` that works both with JSON and BSON.
+///
+/// We only support extracting top-level keys here, so an invocation would look like
+/// `powersync_extract('{"foo": "bar"}', 'foo')`.
+fn powersync_extract_impl(
```

Comment:
Just an early comment here: If you have a typical PowerSync table with say 50 columns, that would mean 50x `powersync_extract` calls for each row in the view, which could mean re-parsing the same JSON/BSON 50 times.

I believe the built-in SQLite json support uses some form of caching of parsed json, which could have a significant impact here. I think it relies on some undocumented behavior though.

I've never tested the performance of this, but it could have quite a big impact for tables with large numbers of columns.